tests/unit/package_managers/test_npm.py

```diff
@@ -242,63 +247,228 @@ def test_get_dependencies(
+                            "version": "4.0.0",
+                            "resolved": "file://file-foo",
+                        },
+                    },
+                },
+                {
+                    "foo": {
+                        "version": "1.0.0",
+                        "name": "foo",
+                        "integrity": None,
+                    },
+                    "bar": {
+                        "version": "2.0.0",
+                        "name": "bar",
+                        "integrity": None,
+                    },
+                    "https://registry.foo.org/@yolo/baz/-/baz-0.16.3.tgz": {
+                        "version": "0.16.3",
+                        "name": "@yolo/baz",
+                        "integrity": "sha512-YOLO8888",
+                    },
+                    "git+ssh://git@foo.org/foo-namespace/git-repo.git#YOLO1234": {
+                        "version": "2.0.0",
+                        "name": "git-repo",
+                        "integrity": None,
+                    },
+                    "https://gitfoo.com/https-namespace/https-tgz/raw/tarball/https-tgz-3.0.0.tgz": {
+                        "version": "3.0.0",
+                        "name": "https-tgz",
+                        "integrity": "sha512-YOLO-4321",
+                    },
+                },
+                id="get_dependencies",
+            ),
+        ],
+    )
+    def test_get_dependencies_to_download(
+        self,
+        rooted_tmp_path: RootedPath,
+        lockfile_data: dict[str, Any],
+        expected_result: bool,
+    ) -> None:
+        package_lock = PackageLock(rooted_tmp_path, lockfile_data)
+        assert package_lock.get_dependencies_to_download() == expected_result
+
+    @pytest.mark.parametrize(
+        "lockfile_data, expected_packages, expected_workspaces",
         [
             pytest.param(
                 {},
                 [],
+                [],
                 id="no_packages",
             ),
             pytest.param(
                 {
                     "packages": {
-                        "": {"version": "1.0.0"},
-                        "node_modules/foo": {"version": "1.0.0"},
-                        "node_modules/bar": {"version": "2.0.0"},
+                        "": {"workspaces": ["foo"], "version": "1.0.0"},
+                        "node_modules/foo": {"version": "1.0.0", "resolved": "foo"},
+                        "node_modules/bar": {"version": "2.0.0", "resolved": "bar"},
                     }
                 },
                 [
-                    Package("foo", "node_modules/foo", {"version": "1.0.0"}),
-                    Package("bar", "node_modules/bar", {"version": "2.0.0"}),
+                    Package("foo", "node_modules/foo", {"version": "1.0.0", "resolved": "foo"}),
+                    Package("bar", "node_modules/bar", {"version": "2.0.0", "resolved": "bar"}),
                 ],
+                [],
```

Comment:
Why are we not expecting any workspaces here? :thinking: 

Edit: I think it's because the packages have `resolved: <filepath>` but don't have `link: true`. I'm not sure what such packages are meant to represent

Reply:
Valid workspace currently is the one which is in the list of workspaces in the main package ("") and at the same time when it has link: True.

Reply:
ack added comment for test cases